tests/integration/single_table/test_copulas.py

```diff
@@ -0,0 +1,83 @@
+from pathlib import Path
+from tempfile import TemporaryDirectory
+
+import pandas as pd
+
+from sdv.datasets.demo import download_demo
+from sdv.single_table import GaussianCopulaSynthesizer
+
+
+def test_adding_constraints():
+    """End to end test for adding constraints to a ``BaseSingleTableSynthesizer``.
+
+    The following functionalities are being tested:
+        * Use an ``Inequality`` constraint.
+        * Load custom constraint class from a file.
+        * Add a custom constraint class to the model.
+        * Validate that the custom constraint was applied properly.
+        * Save, load and sample from the model storing both custom and pre-defined constraints.
+    """
+    # Setup
+    real_data, metadata = download_demo(
+        modality='single_table',
+        dataset_name='fake_hotel_guests'
+    )
+
+    checkin_lessthan_checkout = {
+        'constraint_class': 'Inequality',
+        'constraint_parameters': {
+            'low_column_name': 'checkin_date',
+            'high_column_name': 'checkout_date'
+        }
+    }
+    synthesizer = GaussianCopulaSynthesizer(metadata)
+
+    # Run
+    synthesizer.add_constraints([checkin_lessthan_checkout])
+    synthesizer.fit(real_data)
+    synthetic_data_constrained = synthesizer.sample(500)
+
+    # Assert
+    synthetic_dates = synthetic_data_constrained[['checkin_date', 'checkout_date']].dropna()
+    checkin_dates = pd.to_datetime(synthetic_dates['checkin_date'])
+    checkout_dates = pd.to_datetime(synthetic_dates['checkout_date'])
+    violations = checkin_dates >= checkout_dates
+    assert all(~violations)
+
+    # Load custom constraint class
+    synthesizer.load_custom_constraint_classes(
+        'tests/integration/single_table/custom_constraints.py',
+        ['IfTrueThenZero']
+    )
+    rewards_member_no_fee = {
+        'constraint_class': 'IfTrueThenZero',
+        'constraint_parameters': {
+            'column_names': ['has_rewards', 'amenities_fee']
+        }
+    }
+    synthesizer.add_constraints([rewards_member_no_fee])
+
+    # Re-Fit the model
+    synthesizer.fit(real_data)
+    synthetic_data_custom_constraint = synthesizer.sample(500)
+
+    # Assert
+    validation = synthetic_data_custom_constraint[synthetic_data_custom_constraint['has_rewards']]
+    assert validation['amenities_fee'].sum() == 0.0
+
+    # Save and Load
+    temp_dir = TemporaryDirectory()
```

Comment:
minor: but I see in some places we use `temp_dir = TemporaryDirectory()` and in other tests we pass the `tempdir` as a parameter on the test. Is there a reason to use one over the other and if not should we unify these and just use one throughout our tests?

Reply:
`tempdir` as a parameter is a fixture that comes with `pytest`. I think it's pretty much the same on how we would like to proceed with this. Only pro I see is one less import but a little bit confusing for newcomers if they don't know where the `tmpdir` came from, other than that there is no strong preferences.